core/src/message_queue.c

```diff
@@ -24,15 +24,20 @@ typedef struct MESSAGE_QUEUE_TAG
 static MESSAGE_HANDLE message_pop(MESSAGE_QUEUE_HANDLE_DATA* handle)
 {
     MESSAGE_HANDLE result;
-    MESSAGE_QUEUE_STORAGE* entry = 
+	if (DList_IsListEmpty((PDLIST_ENTRY)&(handle->queue_head)))
+	{
+        /*Codes_SRS_MESSAGE_QUEUE_17_013: [ MESSAGE_QUEUE_pop shall return NULL on an empty message queue. ]*/
+		result = NULL;
+	}
+	else
+	{
+        /*Codes_SRS_MESSAGE_QUEUE_17_014: [ MESSAGE_QUEUE_pop shall remove messages from the queue in a first-in-first-out order. ]*/
+        /*Codes_SRS_MESSAGE_QUEUE_17_015: [ A successful call to MESSAGE_QUEUE_pop on a queue with one message will cause the message queue to be empty. ]*/
+		MESSAGE_QUEUE_STORAGE* entry =
 		(MESSAGE_QUEUE_STORAGE*)DList_RemoveHeadList( (PDLIST_ENTRY)&(handle->queue_head));
-    if (entry == &(handle->queue_head))
-    {
-        result = NULL;
-    }
-    else
-    {
+
         result = ((MESSAGE_QUEUE_STORAGE*)entry)->message;
```

Comment:
DList_RemoveHeadList returns the list head if it's empty, so it should never return NULL.